### ISSUE TITLE
TLS connection issues

### ISSUE THREAD

**Issue by braindev** [ruby-amqp/bunny#303]
I followed the rabbitmq SSL setup here: https://www.rabbitmq.com/ssl.html

When I connect with a MRI ruby I get an error like this:

```
/usr/local/opt/rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/bunny-1.7.0/lib/bunny/transport.rb:81:in `connect': Connection reset by peer - SSL_connect (Errno::ECONNRESET)
	from /usr/local/opt/rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/bunny-1.7.0/lib/bunny/transport.rb:81:in `connect'
	from /usr/local/opt/rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/bunny-1.7.0/lib/bunny/session.rb:273:in `start'
	from test.rb:23:in `<main>'
```

with a corresponding error in the rabbitmq logs:

```
Error on AMQP connection <0.3808.0>:
{ssl_upgrade_failure,
    {{function_clause,
         [{tls_v1,enum_to_oid,[28],[{file,"tls_v1.erl"},{line,404}]},
          {ssl_handshake,'-dec_hello_extensions/2-blc$^1/1-0-',1,
              [{file,"ssl_handshake.erl"},{line,1653}]},
          {ssl_handshake,'-dec_hello_extensions/2-blc$^1/1-0-',1,
              [{file,"ssl_handshake.erl"},{line,1653}]},
          {ssl_handshake,dec_hello_extensions,2,
              [{file,"ssl_handshake.erl"},{line,1653}]},
          {tls_handshake,decode_handshake,3,
              [{file,"tls_handshake.erl"},{line,182}]},
          {tls_handshake,get_tls_handshake_aux,3,
              [{file,"tls_handshake.erl"},{line,153}]},
          {tls_connection,next_state,4,
              [{file,"tls_connection.erl"},{line,454}]},
          {gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]}]},
     {gen_fsm,sync_send_all_state_event,[<0.3809.0>,{start,5000},infinity]}}}
```

However if I use jruby everything works just fine with the same code.  Another interesting thing is the AMQP gem works fine.  I would prefer to use the bunny gem.  Does this sound like a mistake on my part or a bug in the bunny gem?

the bunny code:

```ruby
require 'bunny'

params = {
  :tls                 => true,
  :tls_cert            => "cert.pem",
  :tls_key             => "key.pem",
  :tls_ca_certificates => ["cacert.pem"],
  :verify_peer         => false, # convenient for dev/QA, please enable in production
  :port                => 5671,
  :host                => '192.168.33.10',
  :username            => 'admin',
  #:vhost               => 'broker',
  :password            => '...'

}
conn = Bunny.new(params)

#conn = Bunny.new(vhost: 'broker', username: 'admin', password: '...', host: '192.168.33.10')

conn.start

# open a channel
ch = conn.create_channel

# declare a queue
q  = ch.queue("test1")

# publish a message to the default exchange which then gets routed to this queue
q.publish("Hello, everybody!")

# fetch a message from the queue
delivery_info, metadata, payload = q.pop

puts "This is the message: #{payload}"

# close the connection
conn.stop

puts "done!"
```

the amqp code:

```ruby
require 'amqp'

params = {
  :port                => 5671,
  :ssl => {
    :cert_chain_file  => 'cert.pem',
    :private_key_file => 'key.pem'
  },
  :host                => '192.168.33.10',
  :username            => 'admin',
  :vhost               => 'broker',
  :password            => '...'

}

EventMachine.run do
  connection = AMQP.connect(params)
  puts "Connecting to RabbitMQ. Running #{AMQP::VERSION} version of the gem..."

  ch  = AMQP::Channel.new(connection)
  q   = ch.queue("amqpgem.examples.hello_world", :auto_delete => true)
  x   = ch.default_exchange

  q.subscribe do |metadata, payload|
    puts "Received a message: #{payload}. Disconnecting..."

    connection.close {
      EventMachine.stop { exit }
    }
  end

  x.publish "Hello, world!", :routing_key => q.name
end
```

**Comment by michaelklishin** [comment]
```
```

suggests it is a known issue in Erlang R16* series:

 * http://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2014-April/035637.html
 * http://erlang-questions.erlang.narkive.com/ce22cPw0/making-ssl-fail-silently

March Hare uses a different TLS implementation and may still use `SSLv3` by default (at least in earlier versions), while CRuby uses OpenSSL and Bunny defaults to "TLS 1.0-1.2, whichever is available".

Please use mailing lists for questions in the future.

**Comment by braindev** [comment]
I'm sorry for not using the mailing list.  I should have looked more carefully.

If you don't mind I'll leave a little more I've discovered in case it helps someone in the future.  The ruby version I was using is MRI 2.2.0 compiled against OpenSSL 1.0.2a.  I wrote a program that loops through all the SSL implementations using `OpenSSL::SSL::SSLContext::METHODS` and tried out the rubies I have installed.  I tried MRI 1.9.3 and things work fine. Anything > 1.9.3 does not work.

The really perplexing part is that in a last ditch effort I tried running the program I wrote on linux.  Up to this point only the rabbitmq server has been on linux and the client in MacOSX 10.10. Using ruby 2.2.0 compiled against OpenSSL 1.0.1f on linux everything works fine.
